core/src/main/java/org/apache/calcite/rel/rules/HyperGraph.java

```diff
@@ -0,0 +1,459 @@
+        && !LongBitmap.isOverlap(edge.getLeftNodeBitmap(), subset);
+    return isLeftEnd || isRightEnd;
+  }
+
+  public @Nullable JoinRelType extractJoinType(List<HyperEdge> edges) {
+    JoinRelType joinType = edges.get(0).getJoinType();
+    for (int i = 1; i < edges.size(); i++) {
+      if (edges.get(i).getJoinType() != joinType) {
+        return null;
+      }
+    }
+    return joinType;
+  }
+
+  public RexNode extractJoinCond(RelNode left, RelNode right, List<HyperEdge> edges) {
+    List<RexNode> joinConds = new ArrayList<>();
+    List<RelDataTypeField> fieldList = new ArrayList<>(left.getRowType().getFieldList());
+    fieldList.addAll(right.getRowType().getFieldList());
+
+    List<String> names = new ArrayList<>(left.getRowType().getFieldNames());
+    names.addAll(right.getRowType().getFieldNames());
+
+    // convert the HyperEdge's condition from RexInputFieldName to RexInputRef
+    RexShuttle inputName2InputRefShuttle = new RexShuttle() {
+      @Override protected List<RexNode> visitList(
+          List<? extends RexNode> exprs,
+          boolean @Nullable [] update) {
+        ImmutableList.Builder<RexNode> clonedOperands = ImmutableList.builder();
+        for (RexNode operand : exprs) {
+          RexNode clonedOperand;
+          if (operand instanceof RexInputFieldName) {
+            int index = names.indexOf(((RexInputFieldName) operand).getName());
+            clonedOperand = new RexInputRef(index, fieldList.get(index).getType());
+          } else {
+            clonedOperand = operand.accept(this);
+          }
+          if ((clonedOperand != operand) && (update != null)) {
+            update[0] = true;
+          }
+          clonedOperands.add(clonedOperand);
+        }
+        return clonedOperands.build();
+      }
+    };
+
+    for (HyperEdge edge : edges) {
+      RexNode inputRefCond = edge.getCondition().accept(inputName2InputRefShuttle);
+      joinConds.add(inputRefCond);
+    }
+    return RexUtil.composeConjunction(left.getCluster().getRexBuilder(), joinConds);
+  }
+
+  /**
+   * Before starting enumeration, add Project on every input, make all field name unique.
+   * Convert the HyperEdge condition from RexInputRef to RexInputFieldName
+   */
+  public void convertHyperEdgeCond() {
+    int fieldIndex = 0;
+    List<RelDataTypeField> fieldList = rowType.getFieldList();
+    for (int nodeIndex = 0; nodeIndex < inputs.size(); nodeIndex++) {
+      RelNode input = inputs.get(nodeIndex);
+      List<RexNode> projects = new ArrayList<>();
+      List<String> names = new ArrayList<>();
+      for (int i = 0; i < input.getRowType().getFieldCount(); i++) {
+        projects.add(
+            new RexInputRef(
+            i,
+            fieldList.get(fieldIndex).getType()));
+        names.add(fieldList.get(fieldIndex).getName());
+        fieldIndex++;
+      }
+      RelNode renameProject =
+          LogicalProject.create(
```

Comment:
To make things more generic, I think this method should receive a RelBuilder, and instead of creating a new `LogicalProject`, it should do a relBuilder.push(input) + relBuilder.project(...) + relBuilder.build() to construct the new inputs; that would allow users that do not work with logical-convention to get the appropriate Project operator if the use their own RelBuilder